fms/models/gpt_bigcode.py

```diff
@@ -0,0 +1,331 @@
+
+        self.attn = MultiHeadAttention(
+            self.config.emb_dim,
+            emb_kq,
+            emb_v,
+            self.config.nheads,
+            kvheads,
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        self.ff_sub_layer = FeedForwardBlock(
+            self.config.emb_dim,
+            hidden_grow_factor=self.config.hidden_grow_factor,
+            activation_fn=str_to_activation(self.config.activation_fn),
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        *,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value_state: Optional[
+            Tuple[
+                torch.FloatTensor,
+            ]
+        ] = None,
+        use_cache: bool = False,
+        is_causal_mask: bool = False,
+        attn_algorithm: Optional[str] = None,
+    ):
+
+        self_attn_past_key_value = past_key_value_state
+
+        # first we do MHA and Add&Norm
+        residual = x
+        x = self.ln(x)
+        # self attention
+        x = self.attn(
+            q=x,
+            k=x,
+            v=x,
+            mask=mask,
+            position_ids=position_ids,
+            attn_algorithm=attn_algorithm,
+            past_key_value_state=self_attn_past_key_value,
+            use_cache=use_cache,
+            is_self=True,
+            is_causal_mask=is_causal_mask,
+        )
+
+        cache = None
+        if use_cache:
+            x, cache = x
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # residual connection
+        x = x + residual
+
+        # then we do FF and Add&Norm
+        residual = x
+        x = self.ff_ln(x)
+        x = self.ff_sub_layer(x)
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # another residual
+        x = x + residual
+
+        if use_cache:
+            return x, cache
+        else:
+            return x
+
+
+class GPTBigCodeHeadless(nn.Module):
+    def __init__(self, config: GPTBigCodeConfig):
+        super().__init__()
+        self.config = config
+
+        self.layers = nn.ModuleList(
+            [GPTBigCodeBlock(self.config) for _ in range(self.config.nlayers)]
+        )
+
+        self.embedding = nn.Embedding(self.config.src_vocab_size, self.config.emb_dim)
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+
+        self.dec_norm = nn.LayerNorm(self.config.emb_dim, eps=self.config.ln_eps)
+
+        if self.config.emb_dropout:
+            self.emb_dropout = nn.Dropout(self.config.emb_dropout)
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def _compute_position_ids(
```

Comment:
I think this is much clearer, and the tests make it easier to understand what cases this is handling too.

it might be nice to move this outside the class (maybe positions.py?) so it could be used in other models where it would be relevant.

Reply:
Yes I agree, will add this as an issue to add for later, in case we are missing cases here.